Add action widgets, open page widgets, and gauges on the Lock Screen

literally-anything opened the pull request:
<!-- Thank you for submitting a Pull Request and helping to improve Home Assistant. Please complete the following sections to help the processing and review of your changes. Please do not delete anything from this template. -->

## Summary
Allow action widget, open page widget, and watchOS style complications to be available on the iOS Lock Screen.

## Screenshots
<img width="1163" alt="Screenshot 2024-06-30 at 9 28 19 AM" src="https://github.com/home-assistant/iOS/assets/62899372/be0fd76e-ae4a-494d-b564-9e467533087e">
<img width="722" alt="Screenshot 2024-06-30 at 11 17 13 AM" src="https://github.com/home-assistant/iOS/assets/62899372/d9ac98c8-acf4-4001-9a0c-461ba794c0ee">
<img width="677" alt="Screenshot 2024-06-30 at 9 29 26 AM" src="https://github.com/home-assistant/iOS/assets/62899372/6c2be522-dd78-4325-93e5-29e75ec20576">


## Link to pull request in Documentation repository
<!-- Pull requests that add, change or remove functionality must have a corresponding pull request in the Companion App Documentation repository (https://github.com/home-assistant/companion.home-assistant). Please add the number of this pull request after the "#" -->
Documentation: home-assistant/companion.home-assistant#

## Any other notes
<!-- If there is any other information of note, like if this Pull Request is part of a bigger change, please include it here. -->

bgoncal commented:
Looking good! I will manually test and review the PR this week

bgoncal commented:
btw `bundle exec fastlane autocorrect` to fix lint issues

bgoncal commented on Sources/Extensions/AppIntents/Widget/Details/WidgetDetailsAppIntent.swift:
```diff
@@ -0,0 +1,51 @@
+    static let title: LocalizedStringResource = "Details"
+    static let description = IntentDescription("Display states using from Home Assistant in text")
+    
+    @Parameter(title: "Server", default: nil)
```
Please localize all parameter titles in this file

literally-anything replied:
I can’t figure out localized string resources, I’ll try again tomorrow.

bgoncal replied:
After adding strings to Localizable.strings and building the app, swiftgen will create a wrapper around it that you can access by using `L10n.something.something`

literally-anything replied:
Yeah, I saw that elsewhere in the app, but the app intent parameters seem to want LocalizedStringResource instances. I can only seem to convert a string literal to one, not a string. They seem to just get the string from the strings files themselves using the key, but I can't get them to work, so I might just be missing something.

bgoncal replied:
Hmm interesting, I'll take a look tomorrow and get back to you

literally-anything replied:
Figured it out. The strings file needed to be in the app target for it to find the keys.

bgoncal replied:
If this is the only issue, `LocalizedStringResource` has an option to specify the bundle, which would avoid the need to making Localizable.string available to the main target directly.

But while testing I couldn't make it work either way

bgoncal commented on Sources/Extensions/AppIntents/Widget/Details/WidgetDetailsAppIntentTimelineProvider.swift:
```diff
@@ -0,0 +1,136 @@
+        .init(value: 2, unit: .hours)
+    }
+    
+    static var fastExpiration: Measurement<UnitDuration> {
```
perhaps for this kind of widget that presents info templates we could use the minimum expiration allowed by Apple, I think it is 15 minutes right? (even though the OS may decide to wait for more time of course)

literally-anything replied:
Yeah, I wasn't sure, because of the daily limit

bgoncal replied:
Each widget has a budget system between 40-70 credits, let's try using 15 minutes (the lowerst possible) and then we increase it later if needed

bgoncal commented on Sources/Extensions/AppIntents/Widget/Gauge/WidgetGaugeAppIntent.swift:
```diff
@@ -0,0 +1,104 @@
+    static let title: LocalizedStringResource = "Gauge"
+    static let description = IntentDescription("Display numeric states from Home Assistant in a gauge")
+    
+    @Parameter(title: "Gauge Type", default: .normal)
```
Please localize those titles

bgoncal commented on Sources/Extensions/AppIntents/Widget/Gauge/WidgetGaugeAppIntent.swift:
```diff
@@ -0,0 +1,104 @@
+                    \.$valueTemplate
+                    
+                    \.$valueLabelTemplate
+                    \.$maxTemplate
```
I'd say it makes more sense to first ask for min value then max value

bgoncal commented on Sources/Extensions/AppIntents/Widget/Gauge/WidgetGaugeAppIntent.swift:
```diff
@@ -0,0 +1,104 @@
+    
+    static let typeDisplayRepresentation = TypeDisplayRepresentation(name: "GaugeType")
+    static var caseDisplayRepresentations: [GaugeTypeAppEnum: DisplayRepresentation] = [
+        .normal: DisplayRepresentation(title: "Normal"),
```
Localization missing

bgoncal commented on Sources/Extensions/Widgets/Actions/WidgetActions.swift:
```diff
@@ -32,7 +32,7 @@ struct WidgetActions: Widget {
         .configurationDisplayName(L10n.Widgets.Actions.title)
         .description(L10n.Widgets.Actions.description)
-        .supportedFamilies(WidgetActionSupportedFamilies.families)
+        .supportedFamilies([.systemSmall, .systemMedium, .systemLarge, .systemExtraLarge, .accessoryCircular])
```
Can you keep the enum and than make 2 variables, one for legacy action widgets families and one for the latest action widgets?
And more for other widgets if needed of course

bgoncal commented on Sources/Extensions/Widgets/Common/WidgetBasicContainerView.swift:
```diff
@@ -170,4 +174,14 @@ struct WidgetBasicContainerView: View {
         }
     }
+    
+    private static var clearFamilies: [WidgetFamily] {
```
This variable is not very descriptive in my opinion could you try to rename or make a comment explaining it's purpose?

bgoncal commented on Sources/Extensions/Widgets/Gauge/WidgetGauge.swift:
```diff
@@ -0,0 +1,36 @@
+struct WidgetGauge: Widget {
+    var body: some WidgetConfiguration {
+        AppIntentConfiguration(
+            kind: "io.robbie.HomeAssistant.widget-gauge",
```
Please use an identifier such as what we use in `WidgetActionsIntent.widgetKind'

Just a heads up that the app bundle is dynamic and not always `io.robbie.HomeAssistant`

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -19,6 +20,9 @@ public class NotificationCommandManager {
     public init() {
         register(command: "request_location_update", handler: HandlerLocationUpdate())
         register(command: "clear_notification", handler: HandlerClearNotification())
+        if #available(iOS 14, watchOS 9, macOS 11, *) {
```
iOS 15 and macOS 12 is our minimum requirements, so this can be updated

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -94,6 +98,20 @@ private struct HandlerClearNotification: NotificationCommandHandler {
+        Current.Log.verbose("reloading widgets")
+        return Promise<Void> { seal in
+            DispatchQueue.main.async {
+                WidgetCenter.shared.reloadTimelines(ofKind: "io.robbie.HomeAssistant.widget-gauge")
```
Please update these "kinds" as my previous comment

bgoncal reviewed (changes requested):
Great job! Just some general improvements that I could spot

bgoncal commented:
You also need this code on line 81 in `LegacyNotificationParserImpl`, if you have time would be nice to make it all an enum

```
case "reload_widgets":
                return .init("reload_widgets")
```

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -94,6 +98,20 @@ private struct HandlerClearNotification: NotificationCommandHandler {
+@available(iOS 14, watchOS 9, macOS 11, *)
+private struct HandlerReloadWidgets: NotificationCommandHandler {
+    func handle(_ payload: [String: Any]) -> Promise<Void> {
+        Current.Log.verbose("reloading widgets")
```
```suggestion
        Current.Log.verbose("Reloadings widgets triggered by notification command")
```

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -104,8 +104,8 @@ private struct HandlerReloadWidgets: NotificationCommandHandler {
             DispatchQueue.main.async {
-                WidgetCenter.shared.reloadTimelines(ofKind: "io.robbie.HomeAssistant.widget-gauge")
-                WidgetCenter.shared.reloadTimelines(ofKind: "io.robbie.HomeAssistant.widget-details")
+                WidgetCenter.shared.reloadTimelines(ofKind: "WidgetGauge")
```
Can you make a variable/enum or just something to keep changes aligned between Widget Gauge, WidgetDetails and NotificationsCommandManager?

Just so that if we change the string in one of this places the other one will also get updated automatically.

literally-anything commented:
I haven’t managed to test the notification commands at all. I can’t seem to get my own firebase app connected to APNs properly.

bgoncal commented on Sources/PushServer/SharedPush/Sources/NotificationParserLegacy.swift:
```diff
@@ -246,6 +248,15 @@ public struct LegacyNotificationParserImpl: LegacyNotificationParser {
+    case clearBadge = "clear_badge"
+    case clearNotification = "clear_notification"
+    case updateComplications = "update_complications"
+    case updateWidgets = "update_wigets"
```
Typo 
```suggestion
    case updateWidgets = "update_widgets"
```

bgoncal commented:
> I haven’t managed to test the notification commands at all. I can’t seem to get my own firebase app connected to APNs properly.

I just manually tested it, and after the changes related to the enum above, all works well, we just need to add a note in the documentation telling the user that they can't keep reloading the widgets in a short period of time, since iOS will refuse to do it that often.

Btw, after these changes above we are good to merge, could you just update the docs?
https://companion.home-assistant.io/docs/notifications/notification-commands

repo:
https://github.com/home-assistant/companion.home-assistant

bgoncal commented:
Localization didn't work on my side, do you see it working fine when you edit the widget?

bgoncal commented:
> I haven’t managed to test the notification commands at all. I can’t seem to get my own firebase app connected to APNs properly.

I just manually tested it, and after the changes related to the enum above, all works well, we just need to add a note in the documentation telling the user that they can't keep reloading the widgets in a short period of time, since iOS will refuse to do it that often.

Btw, after these changes above we are good to merge, could you just update the docs?
https://companion.home-assistant.io/docs/notifications/notification-commands

repo:
https://github.com/home-assistant/companion.home-assistant

literally-anything commented:
Localization was working fine for me until I changed my locale. The localized string resources don’t seem to have a fallback for when the string isn’t in the current locale unless you explicitly set it, so I just added defaults for English.

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -19,6 +20,9 @@ public class NotificationCommandManager {
         register(command: "request_location_update", handler: HandlerLocationUpdate())
         register(command: "clear_notification", handler: HandlerClearNotification())
+        if #available(watchOS 9, *) {
+            register(command: "reload_widgets", handler: HandlerReloadWidgets())
```
This is still incorrect

bgoncal replied:
Also, let's remove the #available there and move this inside the #if os(iOS) on  line 26, since it doesnt make sense to have it for watch anyways

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -94,6 +98,20 @@ private struct HandlerClearNotification: NotificationCommandHandler {
     }
 }
 
+@available(watchOS 9, *)
```
Let's remove this available check and move it under the `#if os(iOS)` on line 115

bgoncal reviewed (approved):
LGTM!

bgoncal commented:
Anything else you want to add? Otherwise I can merge it (Asking because you didnt hit the "ready to review" button)

literally-anything commented:
Oh, no I just forgot to hit the button

bgoncal commented:
> Oh, no I just forgot to hit the button

Thank you for your contribution! Don't forget a small update in https://companion.home-assistant.io/docs/notifications/notification-commands

repo:
https://github.com/home-assistant/companion.home-assistant

dnestico commented:
Sorry for commenting but just wondering if any documentation for these new widgets will be added?
